src/SharedStream.mo

```diff
@@ -72,47 +72,77 @@ module {
 
   };
 
-  public type StreamSub<T> = ([T], Nat) -> async ();
-
-
-  /// usage:
-  /// let stream = SharedStream<Int>(10, 10, func (item) = 1);
-  /// stream.next([1, 2, 3, 4]);
-  /// stream.next([5, 6, 7, 8]);
-  /// stream.next([9, 10, 11, 12]);
-  /// stream.subscribe(func (elements: [Int]): async () {
-  ///   await anotherCanister.appendStream(elements); 
-  /// });
-  /// await* stream.emit(); // will send items 1..9 to `anotherCanister`
-  /// await* stream.emit(); // will send items 10..12 to `anotherCanister`
-  /// await* stream.emit(); // will do nothing, stream clear
-  class SharedStream<T>(
-    maxSize : Nat,
-    weightLimit : Nat,
-    weightFunc : (item : T) -> Nat,
+  /// Usage:
+  ///
+  /// let receiver = StreamReceiver<Int>(
+  ///   func (element: Int, index: Nat): () {
+  ///     ... do your logic with item
+  ///   }
+  /// );
+  ///
+  /// Hook-up receive function in the actor class:
+  /// public shared func onStreamChunk(chunk: [Int], firstIndex: Nat) : async () = async receiver.onChunk(chunk, firstIndex);
+  class StreamReceiver<T>(
+    callback : (T, Nat) -> ()
   ) {
 
-    let queue : TemporaryQueueImpl<T> = TemporaryQueueImpl<T>();
-    var subscribers : List<StreamSub<T>> = null;
+    var expectedNextIndex_ : Nat = 0;
 
-    public func subscribe(callback : StreamSub<T>) {
-      subscribers := ?{ e = callback; var next = subscribers };
+    public func onChunk(chunk : [T], firstIndex : Nat) : () {
+      if (firstIndex != expectedNextIndex_) {
+        Debug.trap("Broken chunk index: " # Nat.toText(firstIndex) # "; expected: " # Nat.toText(expectedNextIndex_));
+      };
+      for (index in chunk.keys()) {
+        callback(chunk[index], index);
+      };
```

Comment:
We should send the stream position, not the index within the chunk. Maybe
```
callback(chunk[index], firstIndex + index);
```
does the job?

Reply:
oops, fixed